packages/foam-vscode/src/services/variable-resolver.test.ts

```diff
@@ -0,0 +1,128 @@
+      # \${AnotherVariable:default_value/(.*)/\${1:/upcase}/}} <-- Unrelated to Foam
+      # $AnotherVariable} <-- Unrelated to Foam
+      # $CURRENT_YEAR-\${CURRENT_MONTH}-$CURRENT_DAY <-- Unrelated to Foam
+    `;
+
+    const expected = [];
+
+    expect(findFoamVariables(input)).toEqual(expected);
+  });
+
+  test('it ignores variable-looking text values', () => {
+    const input = `
+      # \${CURRENT_DATE/.*/\${FOAM_TITLE}/} <-- FOAM_TITLE is not a variable here, but a text in a transform
+      # \${1|one,two,\${FOAM_TITLE}|} <-- FOAM_TITLE is not a variable here, but a text in a choice
+    `;
+
+    const expected = [];
+
+    expect(findFoamVariables(input)).toEqual(expected);
+  });
+});
+
+describe('substituteVariables', () => {
+  test('it handles different Variables with the same name', () => {
+    const input = `
+      # $FOAM_TITLE <-- On its own, using the $VARIABLE syntax
+      # \${FOAM_TITLE} <-- On its own, using the \${VARIABLE} syntax
+      # \${1:\${FOAM_TITLE}} <-- In a placeholder
+      # \${2:\${3:\${FOAM_TITLE}}} <-- In a nested placeholder
+      # \${FOAM_TITLE:default_value} <-- With a default
+      # \${FOAM_TITLE/(.*)/\${1:/upcase}/} <-- With a transform
+    `;
+
+    const foamTitle1 = new Variable('FOAM_TITLE', 9, 20);
+    foamTitle1.children.push(new Text('My new note'));
+
+    const foamTitle2 = new Variable('FOAM_TITLE', 72, 85);
+    foamTitle2.children.push(new Text('My new note'));
+
+    const foamTitle3 = new Variable('FOAM_TITLE', 139, 157);
+    foamTitle3.children.push(new Text('My new note'));
+
+    const foamTitle4 = new Variable('FOAM_TITLE', 187, 210);
+    foamTitle4.children.push(new Text('My new note'));
+
+    const foamTitle5 = new Variable('FOAM_TITLE', 247, 274);
+    foamTitle5.children.push(new Text('My new note'));
+
+    const foamTitle6 = new Variable('FOAM_TITLE', 302, 334);
+    foamTitle6.children.push(new Text('MY NEW NOTE'));
+
+    const variables = [
+      foamTitle1,
+      foamTitle2,
+      foamTitle3,
+      foamTitle4,
+      foamTitle5,
+      foamTitle6,
+    ];
+    const expected = `
+      # My new note <-- On its own, using the $VARIABLE syntax
+      # My new note <-- On its own, using the \${VARIABLE} syntax
+      # My new note <-- In a placeholder
+      # My new note <-- In a nested placeholder
+      # My new note <-- With a default
+      # MY NEW NOTE <-- With a transform
+    `;
+
+    expect(substituteVariables(input, variables)).toEqual(expected);
+  });
+
+  test('it substitutes variables at start/end of a string', () => {
+    const input = `\${FOAM_TITLE}`;
+    const variable = new Variable('FOAM_TITLE', 0, 13);
+    variable.children.push(new Text('My new note'));
+
+    const variables = [variable];
+    const expected = 'My new note';
+
+    expect(substituteVariables(input, variables)).toEqual(expected);
+  });
+
+  test('it ignores Variables that are missing positions', () => {
```

Comment:
I am a bit confused here, what do you mean? 
Is the fact that the position is missing in the variable (line 120) that means it doesn't get expanded, or that the variable is not present in the text (line 115)?
What would happen if line 115 were: `const input = '\${FOAM_DATE_MONTH}\n \${FOAM_TITLE}';`?

Reply:
Upon finishing the review, this is my understanding: a `Variable` without position can be resolved, but cannot be replaced in some text. 

Is this correct?

If so, I reckon the reason is that we need the position to know what to replace (accounting for placeholders, transform, ...), which is too hard to find at text replacement time?

I find that possibly confusing, I am trying to understand how much we can work around it

Reply:
> I am a bit confused here, what do you mean? 
> Is the fact that the position is missing in the variable (line 120) that means it doesn't get expanded, or that the variable is not present in the text (line 115)? 

Both. It isn't present in the text, so doesn't get expanded.
This test is just ensuring that `substituteVariables` can handle Variables that do not have positions (because they aren't in the text).

This can happen when variables are given as `extraVariablesToResolve`. For convenience, the caller of `substituteVariables` doesn't have to know anything about `Variables` and doesn't have to filter these out before calling `substituteVariables`.

> What would happen if line 115 were: `const input = '\${FOAM_DATE_MONTH}\n \${FOAM_TITLE}';`?

Then this case wouldn't have happened. The snippet parser would have found `FOAM_TITLE` and would have given it a position.

> Upon finishing the review, this is my understanding: a Variable without position can be resolved, but cannot be replaced in some text.
> Is this correct?

Yes. A `Variable` without a position isn't actually present in the text, so both cannot and should not have any impact on the resulting text.

> I reckon the reason is that we need the position to know what to replace (accounting for placeholders, transform, ...), which is too hard to find at text replacement time?

Our simple regex that we previously had is not powerful enough to correctly identify the bound of a variable in the text. Now that the snippet parser gives us the start and end positions, we know where to replace.

> I find that possibly confusing, I am trying to understand how much we can work around it

Happy to chat more.